### PR TITLE
feat(model-apps): probe what each persona can actually DO, as that persona

### PR DESCRIPTION
**Pull request by akshaymaloo**
> **Ready to review — #425 has merged**, so this is now based on `main` and is a standalone 3-commit change. It depends on `lib/role-privileges.js`, which #425 delivered.
>
> Rebuilt on main after #425 was **squash-merged**: the branch previously carried #425's commits, which conflicted with the squashed version now on main. Only the three probe commits remain — the model-apps content is byte-identical to what was reviewed.
## The gap this closes

#425 added a `role-privileges` check that proves a deployed security role **holds** the privileges the spec declares. That is a metadata comparison, and it stops there.

Whether the persona can *actually perform* the operation additionally depends on **record ownership, business-unit placement, team membership, sharing, and server-side plug-ins** — none of which appear in `roleprivileges`. So a role can verify perfectly clean and still leave the persona unable to do their job.

`probe-persona.js` executes **real reads as the persona** and reports what actually happens.

## Why this needs no human, and no new auth

Dataverse impersonation. Effective privileges are the **intersection** of caller and target, so a System Administrator driving the probe cannot mask a permission the persona lacks.

Two findings made this far cheaper than expected:

1. **`personas[].assignTo.users[]` already holds `systemuserid` GUIDs** — exactly what the legacy `MSCRMCallerID` header takes. So the common case needs **no directory lookup and no application user**. It upgrades to the documented-preferred `CallerObjectId` when `azureactivedirectoryobjectid` is readable.
2. **The existing transport already carries a per-call header** — verified empirically rather than assumed: `Authorization, Accept, OData-MaxVersion, OData-Version, CallerObjectId`, with auth intact. **No change to the http client.**

Caller needs `prvActOnBehalfOfAnotherUser`, assigned **directly** (a team-inherited grant does not satisfy it). [Docs](https://learn.microsoft.com/en-us/power-apps/developer/data-platform/webapi/impersonate-another-user-web-api)

## Three decisions carry the design

**1. It probes the NEGATIVE direction.** For each persona it reads an entity that *another* persona declares and this one does not. An over-broad role is invisible from the inside — every operation the user tries succeeds — so it can only be found by trying something that *should* fail. `appmodule` is excluded, because the build injects it for every persona and it would fail on every run.

**2. An empty `200` on a negative probe is `inconclusive`, never a pass.** Dataverse answers *"no privilege"* with `403` but *"narrower scope"* with a **filtered `200`**, which is indistinguishable from an authorized read of an empty table. Reporting that as a pass would manufacture confidence in the one direction that matters. Inconclusive results do **not** fail the run — they are genuine unknowns, and failing on them would train the operator to ignore the tool — but they are always counted and listed, so an all-inconclusive run cannot masquerade as clean.

**3. A `WhoAmI` canary runs first.** The dangerous failure is not a `403`, which is loud. It is the header being **accepted and silently ignored**: every probe would then run as the signed-in admin, every allow-probe would pass, and the report would look authoritative while proving nothing. `WhoAmI` returns the *effective* user id, so comparing it against the impersonated id catches that. A `403` there reports the real cause rather than a wall of role findings.

## Safety

- **Read-only by default.** `--allow-mutations` only *plans* write probes; it does not execute them. Writing to someone's environment in order to verify it deserves its own explicit design.
- Every un-runnable probe (no test user, unresolvable entity set, transport error) degrades to **inconclusive**, never to a role failure — otherwise the operator is sent to fix a security role when the real problem is a missing test user.

## Scope limit — stated so results are not over-read

This exercises the **Web API**. It says nothing about UCI navigation, form or control visibility, client-side script, the command bar, layout, or accessibility. **A green run means the data operations are authorized — not that the app works.** That limit is recorded in the script header, `AGENTS.md`, the `CHANGELOG` and the skill entry, not just here.

## Placement

`role-privileges` stays in `verify`; this sits **beside** it as an opt-in step in app-builder Phase 3. They are different questions at different costs: verify is free, binary, prerequisite-free and belongs in the build gate; the probe costs N × M round trips, needs a test principal, and can legitimately answer *"inconclusive"* — which cannot serve a gate that must exit zero or non-zero. Moving `role-privileges` out would restore exactly the hole #425 closes.

Invocation is a **script call, not a skill-to-skill call** — the plugin forbids the latter (`AGENTS.md`), while shared scripts are the intended reuse path.

## Verification

- **1511 plugin tests** (1480 → 1511; **31 new**), 0 fail
- Orchestration lives behind **injected IO** (mirroring `readerFor`/`verifySpec`), so principal resolution, entity-set resolution, caching and error containment are all tested without a network
- Tests cover the load-bearing judgements directly: the empty-`200` inconclusive rule, the silently-ignored-header canary, `403` → missing-privilege message, and that `appmodule` is never probed negatively
- Docs updated per the `AGENTS.md` map: file tree, script section, CHANGELOG, skill entry

### CODE REVIEW

File: plugins/model-apps/scripts/probe-persona.js [hunk 1]
```diff
@@ -0,0 +1,200 @@
...
+  }
+  if (!asUser || asUser.status < 200 || asUser.status >= 300) {
+    return { ok: false, detail: `impersonated WhoAmI failed (status ${asUser && asUser.status})` };
+  }
+  const effectiveId = String((asUser.body && asUser.body.UserId) || '').toLowerCase();
+  if (effectiveId && callerId && effectiveId === callerId) {
+    return { ok: false, detail: `the ${header} header was accepted but IGNORED — every probe would run as the signed-in admin and report false passes` };
+  }
+  return { ok: true, effectiveId };
+}
```
[the remark this answers is not in the document]
Comment (akshaymaloo) [reply plugins/model-apps/scripts/probe-persona.js]: Good catch — this was a real false-fail, fixed in `1aa2da65`.

You are right that "effective user == caller" is not sound evidence. A persona whose test user **is** the signed-in user is a legitimate configuration (a single-account dev environment, or an admin validating their own persona), and the canary hard-failed it, blocking the tool for no reason.

Taking your suggestion: the target's `systemuserid` is now the oracle. It is directly comparable to `WhoAmI().UserId`, so it is checked **first** and is the only positive proof:

```js
// Positive proof: the request really ran as the target. Correct even when the target IS the caller.
if (expected && effectiveId && effectiveId === expected) return { ok: true, effectiveId };
```

Equality with the caller now means "header dropped" only when the target differs. I also added the case you implied but did not name: an effective user that is **neither** the caller nor the target previously passed silently, and every finding would then have described the wrong user — that now fails with `unexpected principal`.

One implementation note on your `CallerObjectId` point. The expected id could not be derived from the header value, because under `CallerObjectId` that value is an Entra object id and is not comparable to a systemuserid. So the systemuserid now travels with the principal itself:

```js
return oid
  ? { header: 'CallerObjectId', value: oid, systemUserId: ... }
  : { header: 'MSCRMCallerID',  value: ..., systemUserId: ... };
```

`assignTo.users[]` always carries the systemuserid, so the canary keeps its oracle on both header paths — there is no configuration where it silently degrades to the old heuristic.

Tests 34 → 37, red-green verified: removing the expected-target check turns all three impersonation tests red, including a new one asserting that impersonating yourself **passes**.

File: plugins/model-apps/scripts/lib/persona-probe.js [hunk 2]
```diff
@@ -0,0 +1,275 @@
...
+        mutating: false,
+        reason: `another persona declares ${entity}; this one does not, so it should not be readable`,
+      });
+    }
+  }
+
+  if (!includeMutations && probes.every((p) => !p.mutating)) {
+    const skipped = [...declaredByPersona.values()].flat().filter((d) => isMutating(d.access)).length;
+    if (skipped > 0) {
+      warnings.push(`${skipped} mutating privilege(s) planned but NOT executed (read-only run; pass --allow-mutations to include them)`);
```
[the remark this answers is not in the document]
Comment (akshaymaloo) [reply plugins/model-apps/scripts/lib/persona-probe.js]: Agreed, and fixed in `1aa2da65`. The wording was inaccurate: in a read-only run these privileges are never turned into probes at all, so "planned but NOT executed" implies a plan existed and something stopped it.

```
- ${skipped} mutating privilege(s) planned but NOT executed (read-only run; pass --allow-mutations to include them)
+ ${skipped} mutating privilege(s) were NOT probed — this is a read-only run, so nothing was
+ verified about create/write/delete/append (pass --allow-mutations to probe them)
```

It now states what happened and, more usefully for an operator reading the report, what was therefore **not** verified. The test asserts the message both says "were NOT probed" and does not contain "planned", so the old phrasing cannot come back.
